File: java/src/main/java/com/purescraps/purehtml/transformers/types/LowerTransformer.java
package com.purescraps.purehtml.transformers.types;

import java.util.List;

import com.purescraps.purehtml.transformers.TransformParams;
import com.purescraps.purehtml.transformers.Transformer;

public class LowerTransformer extends Transformer {
    private List<String> args;

    public LowerTransformer(List<String> args) {
        this.args = args;
    }

    public LowerTransformer() {
        this.args = null;
    }

    public void setArgs(List<String> args) {
        this.args = args;
    }

    public String getName() {
        return "lower";
    }

    @Override
    public Object transform(TransformParams params) {
        Object val = params.getVal();

        if (val == null) {
            return null;
        }

        if (val instanceof String) {
            return ((String) val).toLowerCase();
        }

        throw new IllegalArgumentException("lower: expected string. Got " + val);
    }
}

File: python/purehtml/transformers/types/lower_transformer.py
from abc import ABC
from typing import List

from purehtml.transformers.transform_params import TransformParams
from purehtml.transformers.transformer import Transformer


class LowerTransformer(Transformer, ABC):
    def __init__(self, args: List[str] = None):
        super().__init__()
        self.args = args if args is not None else []

    def set_args(self, args: List[str]):
        self.args = args

    def get_name(self) -> str:
        return "lower"

    def transform(self, params: TransformParams):
        val = params.get_val()

        if val is None:
            return None

        if isinstance(val, str):
            return val.lower()

        raise TypeError(f"lower: expected string. Got {val!r}")
